crates/bevy_render/src/lib.rs

```diff
@@ -141,9 +143,8 @@ impl Plugin for RenderPlugin {
                 compatible_surface: surface.as_ref(),
                 ..Default::default()
             };
-            let (device, queue, adapter_info) = futures_lite::future::block_on(
-                renderer::initialize_renderer(&instance, &options, &request_adapter_options),
-            );
+            let (device, queue, adapter_info) =
+                renderer::initialize_renderer(&instance, &options, &request_adapter_options).await;
```

Comment:
https://github.com/bevyengine/bevy/pull/4913 moves this into a closure called once winit says that the app has resumed rather than immediatly running it when building the plugin. This is necessary for android and would solve all the threading through of async plugin building.

Reply:
I *think* this is not enough in this case. I tried using #4913 and it crashed here: https://github.com/bevyengine/bevy/blob/ab72c8368f542c516a494069752b369ba6c2d131/crates/bevy_render/src/lib.rs#L144-L146
The issue with wasm/WebGPU is not to do it later, it's to be able to wait for it without blocking.

I suppose the WebGPU context takes just long enough that it blocks. Some people reported a similar issue when using WebGL2 not in release mode, if initialising the renderer takes too long it doesn't work in wasm to block for it.

Reply:
WIth #4913 it would be possible to turn just the render initialize functions into async functions (hidden from the user) rather than forcing all plugin builders to be async (right in the face of every user).

> I suppose the WebGPU context takes just long enough that it blocks.

Getting a value out of a javascript promise requires either calling `.then()` on it, or `await` on it. Using a mutex on the main thread like futures_lite does isn't possible in javascript. Only workers are allowed to block.

Reply:
> Using a mutex on the main thread like futures_lite does isn't possible in javascript

This is what Bevy is currently doing, and it works... if it's fast enough

Reply:
No, it works on systems where blocking is allowed (non-web platforms). Or when the future is already resolved (webgl on web as webgl isn't async).

Reply:
It should be possible to call the `initialize_renderer` function and then use `wasm_bindgen_futures` to run the rest of the render initialization after the promise resolves.

Reply:
With #4913 the runner is essentially given responsibility to iterate over a vector of callbacks to handle initializing the render state.

One possibility is that those callbacks could be broken up relatively easily if there's some requirement to not block for too long with Web GPU for something. e.g. the runner could potentially run a single callback at a time and return to the browser mainloop between each callback if that would work to keep it happy.

Or, maybe as an alternative to making add_plugin() async then maybe the render_init callbacks could be made async for this. It would probably be fairly simple for the runner to continue running them as blocking functions via `pollster::block_on` in the general case but then there could be some extra care taken for webgpu to let them complete asynchronously.

Maybe the need to make the render_init functions async could be hidden for the most part from most plugins that don't need async render initialization by having separate APIs for `.add_render_init` and `.add_async_render_init` so only the few things dealing with webgpu would utilize async render init callbacks?